aggregator/list.go

```diff
@@ -0,0 +1,343 @@
+	if minFlushInterval := opts.MinFlushInterval(); flushInterval < minFlushInterval {
+		flushInterval = minFlushInterval
+	}
+
+	encoderPool := opts.BufferedEncoderPool()
+	l := &MetricList{
+		opts:          opts,
+		nowFn:         opts.ClockOptions().NowFn(),
+		log:           opts.InstrumentOptions().Logger(),
+		timeLock:      opts.TimeLock(),
+		maxFlushSize:  opts.MaxFlushSize(),
+		flushFn:       opts.FlushFn(),
+		encoderPool:   encoderPool,
+		resolution:    resolution,
+		flushInterval: flushInterval,
+		aggregations:  list.New(),
+		timer:         time.NewTimer(0),
+		encoder:       msgpack.NewAggregatedEncoder(encoderPool.Get()),
+		doneCh:        make(chan struct{}),
+	}
+	l.encodeFn = l.encoder.EncodeChunkedMetricWithPolicy
+	l.aggMetricFn = l.processAggregatedMetric
+	l.waitForFn = time.After
+
+	// Start ticking
+	if l.flushInterval > 0 {
+		l.wgTick.Add(1)
+		go l.tick()
+	}
+
+	return l
+}
+
+// Close closes the list
+func (l *MetricList) Close() {
+	l.Lock()
+	if l.closed {
+		l.Unlock()
+		return
+	}
+	l.closed = true
+	l.Unlock()
+
+	// Waiting for the ticking goroutine to finish
+	close(l.doneCh)
+	l.wgTick.Wait()
+}
+
+// PushBack adds an element to the list
+// NB(xichen): the container list doesn't provide an API to directly
+// insert a list element, therefore making it impossible to pool the
+// elements and manage their lifetimes. If this becomes an issue,
+// need to switch to a custom type-specific list implementation.
+func (l *MetricList) PushBack(value interface{}) (*list.Element, error) {
+	l.Lock()
+	if l.closed {
+		l.Unlock()
+		return nil, errListClosed
+	}
+	elem := l.aggregations.PushBack(value)
+	l.Unlock()
+	return elem, nil
+}
+
+// tick performs periodic maintenance tasks (e.g., flushing out aggregated metrics)
+func (l *MetricList) tick() {
+	defer l.wgTick.Done()
+
+	for {
+		select {
+		case <-l.doneCh:
+			return
+		default:
+			l.tickInternal()
+		}
+	}
+}
+
+func (l *MetricList) tickInternal() {
+	// NB(xichen): it is important to determine ticking start time within the time lock
+	// because this ensures all the actions before `start` have completed if those actions
+	// are protected by the same read lock
+	l.timeLock.Lock()
+	start := l.nowFn()
+	resolution := l.resolution
+	l.timeLock.Unlock()
+	alignedStart := start.Truncate(resolution)
+
+	// Reset states reused across ticks
+	l.idBuf.Reset()
+	l.toCollect = l.toCollect[:0]
+
+	// Flush out aggregations, may need to do it in batches if the read lock
+	// is held for too long
+	l.RLock()
+	for e := l.aggregations.Front(); e != nil; e = e.Next() {
+		switch elem := e.Value.(type) {
+		case metricElem:
+			// If the element is eligible for collection after the values are
+			// processed, close it and reset the value to nil
+			if elem.Consume(alignedStart, l.aggMetricFn) {
+				elem.Close()
+				e.Value = nil
+				l.toCollect = append(l.toCollect, e)
+			}
+		default:
+			panic(fmt.Errorf("unknown metric element type: %T", elem))
```

Comment:
Seems some places we check the type and panic if not and others we just cast and call directly against the cast.  Perhaps just do one or the other?  I feel we probably should just be casting and calling on it rather than checking type.  We'll panic anyway just in a slightly different way and its less code overall.